allennlp/data/transforms.py

```diff
@@ -0,0 +1,558 @@
+
+        if batch:
+            yield batch
+
+
+@Transform.register("homogenous_batches_of")
+class HomogenousBatchesOf(InstanceTransform[Batched]):
+
+    """
+    This Transform takes a dataset of potentially heterogeneous instances
+    and yields back homogeneous batches. It assumes that each instance has
+    some ``MetadataField`` indicating what "type" of instance it is
+    and bases its notion of homogeneity on that (and, in particular, not on
+    inspecting the "field signature" of the instance.)
+
+    Parameters
+    ----------
+    batch_size : ``int``, required
+        The batch size to use to batch instances.
+    partition_key : ``str``, optional, (default = "dataset")
+        The key of the ``MetadataField`` indicating what "type" of instance this is.
+    in_metadata : ``boool``: optional (default = False)
+        Whether the partition key actually exists in a Field with the name "metadata",
+        rather than being a field itself.
+    """
+
+    def __init__(self, batch_size: int, partition_key: str = "dataset", in_metadata: bool = False):
+
+        self.batch_size = batch_size
+        self.partition_key = partition_key
+        self.in_metadata = in_metadata
+
+    def transform(self, dataset: Iterable[Instance]) -> Iterable[Batched]:
+
+        instances = list(dataset)
+
+        hoppers: Dict[str, List[Instance]] = defaultdict(list)
+
+        for instance in instances:
+            if self.in_metadata:
+                partition = instance.fields["metadata"].metadata[self.partition_key]  # type: ignore
+            else:
+                partition = instance.fields[self.partition_key].metadata  # type: ignore
+            hoppers[partition].append(instance)
+
+        # Get a `lazy_groups_of` iterator over each set of homogeneous instances.
+        batches = {
+            key: lazy_groups_of(iter(hopper), self.batch_size) for key, hopper in hoppers.items()
+        }
+
+        remaining = set(batches)
+        # Yield batches in a round-robin fashion until none are left.
+        while remaining:
+            for key, lazy_batches in batches.items():
+                if key in remaining:
+                    try:
+                        batch = next(lazy_batches)
+                        yield batch
+                    except StopIteration:
+                        remaining.remove(key)
+
+
+@Transform.register("biggest_batch_first")
+class BiggestBatchFirst(InstanceTransform[Batched]):
+    def transform(self, dataset: Iterable[Instance]) -> Iterable[Batched]:
+
+        raise NotImplementedError(
+            "BiggestBatchFirst must be called on batches. "
+            "Add a transform which batches your instances to your pipeline."
+        )
+
+    def transform_batch(self, batches: Iterable[Batched]) -> Iterable[Batched]:
+
+        batches = list(batches)
+        if len(batches) <= 2:
+            yield from batches
+        else:
+            last = batches.pop()
```

Comment:
I don't think this does what's advertised.  Why will the last batches be biggest here?